### PR TITLE
Feature/add history

### PR DESCRIPTION
**Pull request by ndoriomedov**

### CODE REVIEW

**Review by VLiamin — APPROVED** [review]

File: src/ProjectService.Business/Commands/ProjectUsers/CreateProjectUsersCommand.cs [hunk 1]
```diff
@@ -72,36 +73,24 @@ public async Task<OperationResultResponse<bool>> ExecuteAsync(ProjectUsersReques
         return _responseCreator.CreateFailureResponse<bool>(HttpStatusCode.BadRequest, errors);
       }
 
-      List<Guid> existUsers = await _repository.GetExistAsync(request.ProjectId, request.Users.Select(u => u.UserId).ToList());
+      List<DbProjectUser> existingUsers = await _repository.GetExistingUsersAsync(request.ProjectId, request.Users.Select(u => u.UserId));
 
-      request.Users = request.Users.GroupBy(u => u.UserId).Select(g => g.First()).Where(u => !existUsers.Contains(u.UserId)).ToList();
+      request.Users = request.Users.ExceptBy(existingUsers.Select(x => x.UserId), request => request.UserId).ToList();
```
Comment (MariyaSin) [inline src/ProjectService.Business/Commands/ProjectUsers/CreateProjectUsersCommand.cs:78]: i thik it is bad practise to change request

File: src/ProjectService.Business/Commands/ProjectUsers/CreateProjectUsersCommand.cs [hunk 2]
```diff
@@ -72,36 +73,24 @@ public async Task<OperationResultResponse<bool>> ExecuteAsync(ProjectUsersReques
...
-
-      if (existUsers.Any())
-      {
-        errors.Add("Exist user were not added again.");
-      }
+      bool result = await _repository.CreateAsync(_mapper.Map(request), existingUsers);
 
       if (result)
       {
+        await _bus.Publish<ICreateWorkTimePublish>(ICreateWorkTimePublish.CreateObj(
```
Comment (MariyaSin) [inline src/ProjectService.Business/Commands/ProjectUsers/CreateProjectUsersCommand.cs:84]: please create new class

File: src/ProjectService.Data/ProjectUserRepository.cs [hunk 3]
```diff
@@ -78,10 +78,10 @@ public async Task<List<DbProjectUser>> GetAsync(List<Guid> usersIds)
       return await _provider.ProjectsUsers.Where(u => usersIds.Contains(u.UserId) && u.IsActive).ToListAsync();
     }
 
-    public async Task<List<Guid>> GetExistAsync(Guid projectId, IEnumerable<Guid> usersIds)
+    public async Task<List<DbProjectUser>> GetExistingUsersAsync(Guid projectId, IEnumerable<Guid> usersIds)
     {
       return await _provider.ProjectsUsers
-        .Where(pu => pu.IsActive && pu.ProjectId == projectId && usersIds.Contains(pu.UserId)).Select(pu => pu.UserId).ToListAsync();
+        .Where(pu => pu.ProjectId == projectId && usersIds.Contains(pu.UserId)).ToListAsync();
```
Comment (MariyaSin) [inline src/ProjectService.Data/ProjectUserRepository.cs:84]: Should you check if user isActive?

File: src/ProjectService.Data/ProjectUserRepository.cs [hunk 4]
```diff
@@ -103,15 +103,23 @@ public async Task<(List<DbProjectUser>, int totalCount)> GetAsync(IGetProjectsUs
       return (await projectUsers.ToListAsync(), totalCount);
     }
 
-
-    public async Task<bool> CreateAsync(List<DbProjectUser> dbProjectUsers)
+    public async Task<bool> CreateAsync(List<DbProjectUser> newUsers, List<DbProjectUser> oldUsers)
```
Comment (MariyaSin) [inline src/ProjectService.Data/ProjectUserRepository.cs:106]: split create and edit methods

File: src/ProjectService.Data/ProjectUserRepository.cs (hunk 3 above)
Comment (ndoriomedov) [reply src/ProjectService.Data/ProjectUserRepository.cs:84]: no - this method is to check if user is in project now (even with isActive - false) in createProjectUser command

File: src/ProjectService.Broker/Publishes/PublishHelper.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  {
+    private readonly IBus _bus;
+
+    public PublishHelper(
+      IBus bus)
+    {
+      _bus = bus;
+    }
+
+    public async Task CreateDepartmentEntityPublish(Guid departmentId, Guid createdBy, Guid? userId = null, Guid? projectId = null)
```
Comment (MariyaSin) [inline src/ProjectService.Broker/Publishes/PublishHelper.cs:21]: projectId is not nullable

Comment (MariyaSin) [reply src/ProjectService.Broker/Publishes/PublishHelper.cs:21]: remove userId

File: src/ProjectService.Business/Commands/Project/EditProjectCommand.cs [hunk 6]
```diff
@@ -73,12 +73,10 @@ public async Task<OperationResultResponse<bool>> ExecuteAsync(Guid projectId, Js
 
       response.Body = await _projectRepository.EditAsync(projectId, _mapper.Map(request));
 
-      response.Status = OperationResultStatusType.FullSuccess;
       if (!response.Body)
       {
         _httpContextAccessor.HttpContext.Response.StatusCode = (int)HttpStatusCode.BadRequest;
 
-        response.Status = OperationResultStatusType.Failed;
         response.Errors.Add("Project can not be edit.");
```
Comment (MariyaSin) [inline src/ProjectService.Business/Commands/Project/EditProjectCommand.cs:80]: remove

File: src/ProjectService.Validation/User/CreateProjectUsersRequestValidator.cs [hunk 7]
```diff
@@ -7,29 +7,34 @@
...
         .NotEmpty()
         .WithMessage("Request must have a project Id")
-        .MustAsync(async (x, _) => await projectRepository.DoesExistAsync(x))
+        .MustAsync(async (projectId, _) => await projectRepository.DoesExistAsync(projectId))
         .WithMessage("This project id does not exist")
         .DependentRules(() =>
         {
-          RuleFor(projectUser => projectUser.Users)
+          RuleFor(request => request.Users)
             .Must(users => users is not null && users.Any())
```
Comment (MariyaSin) [inline src/ProjectService.Validation/User/CreateProjectUsersRequestValidator.cs:28]: do you need check is not null?

File: src/ProjectService.Validation/User/CreateProjectUsersRequestValidator.cs [hunk 8]
```diff
@@ -7,29 +7,34 @@
...
             .WithMessage("The request must contain users")
             .MustAsync(async (projectUser, cancellation) =>
```
Comment (MariyaSin) [inline src/ProjectService.Validation/User/CreateProjectUsersRequestValidator.cs:30]: rename to user and _

File: src/ProjectService.Mappers/Models/ProjectInfoMapper.cs [hunk 9]
```diff
@@ -7,24 +7,27 @@ namespace LT.DigitalOffice.ProjectService.Mappers.Models
 {
   public class ProjectInfoMapper : IProjectInfoMapper
   {
-    public ProjectInfo Map(DbProject value, DepartmentInfo department)
+    public ProjectInfo Map(DbProject dbProject, DepartmentInfo department)
     {
-      if (value == null)
+      if (dbProject == null)
```
Comment (MariyaSin) [inline src/ProjectService.Mappers/Models/ProjectInfoMapper.cs:12]: use ?:

File: src/ProjectService.Data/ProjectUserRepository.cs [hunk 10]
```diff
@@ -103,21 +83,39 @@ public async Task<(List<DbProjectUser>, int totalCount)> GetAsync(IGetProjectsUs
...
+      await _provider.SaveAsync();
 
-    public async Task<bool> CreateAsync(List<DbProjectUser> dbProjectUsers)
+      return true;
+    }
+
+    public async Task<bool> ReturnUsersAsync(List<DbProjectUser> oldUsers, Guid createdBy)
     {
-      if (dbProjectUsers == null)
+      if (oldUsers is null)
```
Comment (MariyaSin) [inline src/ProjectService.Data/ProjectUserRepository.cs:102]: rename to dbUsers

File: src/ProjectService.Mappers/Models/ProjectInfoMapper.cs (hunk 9 above)
Comment (ndoriomedov) [reply src/ProjectService.Mappers/Models/ProjectInfoMapper.cs:12]: i think now it is clearer

**Pull request merged**